File: julia/ch1/Ex127.jl
"""
```jldoctest
julia> include("c:/Users/Yohan/IdeaProjects/sicp/julia/ch1/ExXXX.jl")

julia> ExXXX.func(arg)

or
shell> julia ExXXX.jl
```
"""
module Ex127

fast_times = 128

sq(n) = n * n

divides(a, b) = b % a == 0

function find_divisor(n, test)
    if sq(test) > n
        return n
    elseif divides(test, n)
        return test
    else
        return find_divisor(n, test + 1)
    end
end

smallest_divisor(n) = find_divisor(n, 2)

prime(n) = n == smallest_divisor(n)

function fast_prime(n, times)
    if times == 0
        return true
    elseif fermat_test(n)
        return fast_prime(n, times-1)
    else
        return false
    end
end 

function carmichael_num(n)
    try_it(a) = a == expmod(a, n, n)
    iter(x) = x == 0 ? true : (try_it(x) ? iter(x-1) : false)
    return !prime(n) && iter(n-1)
end

function expmod(base, exp, m)
    if exp == 0
        return 1
    elseif divides(2, exp)
        return rem(sq(expmod(base, div(exp,2), m)), m)
    else
        return rem(base * expmod(base, exp-1, m), m)
    end
end

function fermat_test(n)
    try_it(a) = expmod(a, n, n) == a
    return try_it(1 + random(n - 1))
end

random(limit) = rem(abs(rand(Int, 1)[1]), limit)

function test_case(n)
    p = prime(n)
    fp = fast_prime(n, fast_times)
    car = carmichael_num(n)
    return "$n \t Prime: $p \t Fast Prime: $fp \t Carmichael: $car"
end

end

# testing
println(Ex127.test_case(10))
println(Ex127.test_case(11))
println(Ex127.test_case(13))
println(Ex127.test_case(97))
println(Ex127.test_case(561))
println(Ex127.test_case(1105))
println(Ex127.test_case(1729))
println(Ex127.test_case(2465))
println(Ex127.test_case(2821))
println(Ex127.test_case(6601))
println(Ex127.test_case(6603))
println("done")

File: ruby/ch1/ex127.rb
module Ex127

def divides(a, b)
  0 == b % a
end

def square(n)
  n * n
end

def find_divisor(n, test)
  if test * test > n
    n
  elsif divides(test, n)
    test
  else
    find_divisor(n, test + 1)
  end
end 

def smallest_divisor(n)
  find_divisor(n, 2)
end

def is_prime(n)
  n == smallest_divisor(n)
end

def fast_prime(n, times)
  if times == 0
    true
  elsif fermat_test(n)
    fast_prime(n, times-1)
  else
    false
  end
end

def fermat_test(n)
  try_it = lambda { |a| a == expmod(a, n, n) }
  try_it.call(1 + rand(n-1))
end

def expmod(base, exp, m)
  if exp == 0
    1
  elsif divides(2, exp)
    square(expmod(base, exp/2, m)) % m
  else
    (base * expmod(base, exp-1, m)) % m
  end
end

def carmichael_num(n)
  def iter(x, n)
    try_it = lambda { |a| a == expmod(a, n, n) }
    if x == 0
      true
    elsif try_it.call(x)
      iter(x-1, n)
    else
      false
    end
  end

  (not is_prime(n)) and iter(n-1, n)
end

def test_case(n)
  prime = is_prime(n)
  fp = fast_prime(n, 4)
  car = carmichael_num(n)
  print("#{n} \t Prime: #{prime} \t Fast Prime:#{fp} \t Carmichael:#{car}\n")
end


end

include Ex127
print "Testing\n"

test_case(9)
test_case(10)
test_case(11)
test_case(71)
test_case(97)
test_case(561)
test_case(1105)
test_case(1729)
test_case(2465)
test_case(2821)
test_case(6601)

print "done", "\n"
